### PR TITLE
Improvement and testing

### PR DESCRIPTION
**Pull request by ritz078**
@pranjal-jain photostory can be developed by extending Swipe in the same repo. It can be exported separately or it can accept swipe as child. Now this supports children so you can make a slideshow of even html elements.

@cgarg @dxtr026 @azizhk @vivekj011

### CODE REVIEW

File: components/Swipe.js [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+            className='rs-imgs-wrapper'
+            style={style}
+            onTouchStart={this.handleTouchStart}
+            onTouchEnd={this.handleTouchEnd}
+            onClick={this.handleClick}
+          >
+            {children$}
+          </div>
+        </div>
+        {!this.hasSingleImage() && <div className={prevClass} onClick={this.gotoPrev}>{prev}</div>}
```
Comment (dxtr026) [inline components/Swipe.js:205]: maybe moving prev before rs-imgs-wrapper gives better control, but if they are absolutely positioned thats ok.

File: components/Swipe.js [hunk 2]
```diff
@@ -0,0 +1,279 @@
+import React, { Component, PropTypes } from 'react'
+import classNames from 'classnames'
+
+function autoBind (methods, context) {
+  methods.forEach(method => {
+    // eslint-disable-next-line no-param-reassign
+    context[method] = context[method].bind(context)
+  })
+}
```
Comment (parthshah1991) [inline components/Swipe.js]: util method, use in all components?

File: components/Swipe.js [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+      'initLazyLoad',
+      'hasSingleImage',
+      'autoPlay',
+      'pause',
+      'handleClick'
+    ], this)
+  }
+
+  componentDidMount () {
+    this.setWidth()
```
Comment (pranjal-jain) [inline components/Swipe.js:19]: https://github.com/facebook/react/issues/2659

File: components/Swipe.js [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+  }
+
+  gotoSlide (i) {
+    const initial = this.state.currentIndex
+    this.setState({
+      currentIndex: i
+    }, () => (this.onChange(initial)))
+  }
+
+  autoPlay (props) {
```
Comment (parthshah1991) [inline components/Swipe.js:121]: Test Case for component with auto play:

When I manually go to the next slide, this.pause() should be called, else after time out, the user will go to the next screen.

File: components/Swipe.js [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+            style={style}
+            onTouchStart={this.handleTouchStart}
+            onTouchEnd={this.handleTouchEnd}
+            onClick={this.handleClick}
+          >
+            {children$}
+          </div>
+        </div>
+        {!this.hasSingleImage() && <div className={prevClass} onClick={this.gotoPrev}>{prev}</div>}
+        {!this.hasSingleImage() && <div className={nextClass} onClick={this.gotoNext}>{next}</div>}
```
Comment (parthshah1991) [inline components/Swipe.js:206]: Can cache the method return value.

File: components/swipe.css [hunk 6]
```diff
@@ -6,9 +6,10 @@
 
 .rs-imgs-wrapper{
   height: 100%;
+  width: 100%;
   overflow: hidden;
   will-change: transform;
-  transition: transform .2s ease-out;
+  transition: transform .3s ease-out;
```
Comment (parthshah1991) [inline components/swipe.css:12]: browser prefixers required

File: components/swipe.css [hunk 7]
```diff
@@ -17,6 +18,7 @@
   background-size: cover;
   float: left;
   opacity: 0.6;
+  width: 100%;
   will-change: opacity;
   transition: opacity .3s;
```
Comment (parthshah1991) [inline components/swipe.css:23]: same

File: components/Swipe.js [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+  }
+
+  componentDidMount () {
+    this.setWidth()
+    this.initLazyLoad()
+
+    if (this.props.autoPlay) this.autoPlay(this.props)
+
+    if (this.props.responsive) {
+      window.addEventListener('resize', this.setWidth, false)
```
Comment (pranjal-jain) [inline components/Swipe.js:25]: Remove in willUnmount

**Review by parthshah1991 — CHANGES_REQUESTED** [review]

**Review by pranjal-jain — CHANGES_REQUESTED** [review]

File: components/Swipe.js [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+  onChange (initialIndex) {
+    this.props.onSwipe({
+      currentIndex: this.state.currentIndex,
+      initialIndex
+    })
+    this.initLazyLoad()
+  }
+
+  hasSingleImage () {
+    return !(this.props.children.length > 1)
```
Comment (vivekj011) [inline components/Swipe.js]: keep length == 1 check

File: components/Swipe.js [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+  }
+
+  handleClick () {
+    this.props.onClick({
+      index: this.state.currentIndex
+    })
+  }
+
+  componentWillUnmount () {
+    window.removeEventListener('resize', this.setWidth)
```
Comment (vivekj011) [inline components/Swipe.js:160]: if (this.props.responsive)

File: tests/Swipe.test.js [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+
+    wrapper.find('.rs-imgs-wrapper').simulate('touchstart', {
+      touches: [{
+        clientX: 0
+      }]
+    })
+
+    wrapper.setState({drag})
+
+    wrapper.find('.rs-imgs-wrapper').simulate('touchend')
```
Comment (vivekj011) [inline tests/Swipe.test.js:31]: behavioural test 👍

**Pull request merged**